src/ERC4626StreamHub.sol

```diff
@@ -0,0 +1,206 @@
+        // equivalent to the sender share of the total principal
+        uint256 shares = ask > have ? have : ask;
+
+        // update state and transfer
+        receiverPrincipal[_receiver][msg.sender] = 0;
+        receiverTotalPrincipal[_receiver] -= totalPrincipal;
+        receiverShares[_receiver] -= shares;
+
+        vault.safeTransfer(msg.sender, shares);
+
+        emit CloseYieldStream(msg.sender, _receiver, shares);
+    }
+
+    /**
+     * @dev Closes multiple yield streams for multiple receivers from the same streamer.
+     * If there is any yield to claim for the stream, it will remain unclaimed until the receiver calls `claimYield` function.
+     * @param _tos The array of receiver addresses.
+     */
+    function closeYieldStreamBatch(address[] calldata _tos) external {
+        for (uint256 i = 0; i < _tos.length; i++) {
+            closeYieldStream(_tos[i]);
+        }
+    }
+
+    /**
+     * @dev Claims the yield for the sender and transfers it to the specified receiver address.
+     * @param _to The address to receive the claimed yield.
+     * @return assets The amount of assets (tokens) claimed as yield.
+     */
+    function claimYield(address _to) external returns (uint256 assets) {
+        _checkZeroAddress(_to);
+
+        uint256 principalInShares = _convertToShares(
+            receiverTotalPrincipal[msg.sender]
+        );
+        uint256 shares = receiverShares[msg.sender];
+
+        // if vault made a loss, there is no yield to claim
+        if (shares <= principalInShares) revert NoYieldToClaim();
+
+        uint256 yieldShares = shares - principalInShares;
+
+        receiverShares[msg.sender] -= yieldShares;
+
+        assets = vault.redeem(yieldShares, _to, address(this));
+
+        emit ClaimYield(msg.sender, _to, assets);
+    }
+
+    /**
+     * @dev Calculates the yield for a given receiver.
+     * @param _receiver The address of the receiver.
+     * @return The calculated yield.
+     */
+    function yieldFor(address _receiver) public view returns (uint256) {
+        uint256 principal = receiverTotalPrincipal[_receiver];
+        uint256 currentValue = _convertToAssets(receiverShares[_receiver]);
+
+        return currentValue > principal ? currentValue - principal : 0;
+    }
+
+    function _checkZeroAddress(address _receiver) internal pure {
+        if (_receiver == address(0)) revert AddressZero();
+    }
+
+    function _checkZeroShares(uint256 _shares) internal pure {
+        if (_shares == 0) revert ZeroShares();
+    }
+
+    function _checkReceiver(address _receiver) internal view {
+        if (_receiver == msg.sender) revert CannotOpenStreamToSelf();
+    }
+
+    function _convertToAssets(uint256 _shares) internal view returns (uint256) {
+        return vault.convertToAssets(_shares);
+    }
+
+    function _convertToShares(uint256 _assets) internal view returns (uint256) {
+        return vault.convertToShares(_assets);
+    }
```

Comment:
maybe a bit, but `vault.convertTo...` are external calls used in the same way in a few places so I decided to have it as internal functions since it looked cleaner and it reduces the contract size